File: tests/fixtures-js/src/models/player.js
// The annotated source the JavaScript integration tests exercise - the
// JavaScript counterpart of tests/fixtures-ts/src/models/player.ts, with
// every type annotation dropped: the same annotations mean the same thing
// in both languages (issue.md §9).

// A field whose network type is another model.
// FOMOXA_MODEL
// FOMOXA_CODEC("edge")
export class PlayerInfo {
    // FOMOXA_FIELD(u32)
    // FOMOXA_CODEC("edge")
    Level = 0;
}

// The model RFC-0002 §9.1 is tested against: three fields, and a version
// that appends a fourth.
// FOMOXA_MODEL
// FOMOXA_CODEC("edge", "unity")
export class Player {
    // FOMOXA_FIELD(u32)
    // FOMOXA_CODEC("edge", "unity")
    Id = 0;

    // FOMOXA_FIELD(f32)
    // FOMOXA_CODEC("edge")
    X = 0;

    // FOMOXA_FIELD(f32)
    // FOMOXA_CODEC("edge")
    Y = 0;

    // A network field in no codec: it is written by none of them.
    // FOMOXA_FIELD(u32)
    Unrouted = 0;

    // Not a network field at all. Logic and caches stay off the wire.
    Cache = "";
}

// Composites: an array of primitives, an array of models, and a nested
// model.
// FOMOXA_MODEL
// FOMOXA_CODEC("edge")
export class Team {
    // FOMOXA_FIELD(PlayerInfo)
    // FOMOXA_CODEC("edge")
    Captain = new PlayerInfo();

    // FOMOXA_FIELD(Array<string>)
    // FOMOXA_CODEC("edge")
    Tags = [];

    // FOMOXA_FIELD(Array<u32>)
    // FOMOXA_CODEC("edge")
    Scores = [];

    // FOMOXA_FIELD(Array<PlayerInfo>)
    // FOMOXA_CODEC("edge")
    Roster = [];
}

File: tests/fixtures/src/models/player.rs
use fomoxa_attributes::*;

// Nested models and arrays, and the model the version-skew tests decode.

/// A field whose network type is another model.
#[network]
#[codec(edge)]
#[derive(Debug, Default, PartialEq)]
pub struct PlayerInfo {
    #[network(u32)]
    #[codec(edge)]
    pub level: u32,
}

/// The model RFC-0002 §9.1 is tested against: three fields, and a version that
/// appends a fourth.
#[network]
#[codec(edge)]
#[derive(Debug, Default, PartialEq)]
pub struct Player {
    #[network(u32)]
    #[codec(edge)]
    pub id: u32,

    #[network(f32)]
    #[codec(edge)]
    pub x: f32,

    #[network(f32)]
    #[codec(edge)]
    pub y: f32,
}

/// Composites: an array of primitives, an array of models, and a nested model.
#[network]
#[codec(edge)]
#[derive(Debug, Default, PartialEq)]
pub struct Team {
    #[network(PlayerInfo)]
    #[codec(edge)]
    pub captain: PlayerInfo,

    #[network(Array<string>)]
    #[codec(edge)]
    pub tags: Vec<String>,

    #[network(Array<u32>)]
    #[codec(edge)]
    pub scores: Vec<u32>,

    #[network(Array<PlayerInfo>)]
    #[codec(edge)]
    pub roster: Vec<PlayerInfo>,
}
